libkbfs/file_data.go

```diff
@@ -128,6 +132,95 @@ func (fd *fileData) getFileBlockAtOffset(ctx context.Context,
 	return ptr, parentBlocks, block, nextBlockStartOff, startOff, wasDirty, nil
 }
 
+// getNextDirtyFileBlockAtOffset returns the next dirty leaf block
+// with a starting offset that is equal or greater than the given
+// `off`.  This assumes that any code that dirties a leaf block also
+// dirties all of its parents, even if those parents haven't yet
+// changed.  It can be used iteratively (by feeding
+// `nextBlockStartOff` back in as `off`) to find all the dirty blocks
+// of the file.  Note that there is no need to parallelize that
+// process, since all the dirty blocks are guaranteed to be local.
+func (fd *fileData) getNextDirtyFileBlockAtOffset(ctx context.Context,
+	topBlock *FileBlock, off int64, rtype blockReqType,
+	dirtyBcache DirtyBlockCache) (
+	ptr BlockPointer, parentBlocks []parentBlockAndChildIndex,
+	block *FileBlock, nextBlockStartOff, startOff int64,
+	err error) {
+	// Find the block matching the offset, if it exists.
+	ptr = fd.rootBlockPointer()
+	if !dirtyBcache.IsDirty(fd.file.Tlf, ptr, fd.file.Branch) {
+		// The top block isn't dirty, so we know none of the leaves
+		// are dirty.
+		return zeroPtr, nil, nil, 0, 0, nil
+	}
+
+	block = topBlock
+	nextBlockStartOff = -1
+	startOff = 0
+	// Search along paths of dirty blocks until we find a dirty leaf
+	// block with an offset equal or greater than `off`.
+	for block.IsInd {
```

Comment:
I haven't yet gone through this function in detail, but this search-for-condition-among-leaves pattern seems something we implement multiple times. Is there any way to DRY it a bit better?